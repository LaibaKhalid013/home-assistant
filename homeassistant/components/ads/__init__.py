"""Support for Automation Device Specification (ADS)."""

import asyncio
from asyncio import timeout
from collections import namedtuple
import ctypes
from enum import Enum
import logging
import struct
import threading
from typing import Any

import pyads
import pyads.errorcodes
import voluptuous as vol

from homeassistant.const import (
    CONF_DEVICE,
    CONF_IP_ADDRESS,
    CONF_PORT,
    EVENT_HOMEASSISTANT_STOP,
)
from homeassistant.core import HomeAssistant, ServiceCall
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.entity import Entity
from homeassistant.helpers.typing import ConfigType

_LOGGER = logging.getLogger(__name__)

DATA_ADS = "data_ads"

# Supported Types
ADSTYPE_BOOL = "bool"
ADSTYPE_BYTE = "byte"
ADSTYPE_DINT = "dint"
ADSTYPE_INT = "int"
ADSTYPE_UDINT = "udint"
ADSTYPE_UINT = "uint"

ADS_TYPEMAP = {
    ADSTYPE_BOOL: pyads.PLCTYPE_BOOL,
    ADSTYPE_BYTE: pyads.PLCTYPE_BYTE,
    ADSTYPE_DINT: pyads.PLCTYPE_DINT,
    ADSTYPE_INT: pyads.PLCTYPE_INT,
    ADSTYPE_UDINT: pyads.PLCTYPE_UDINT,
    ADSTYPE_UINT: pyads.PLCTYPE_UINT,
}

CONF_ADS_FACTOR = "factor"
CONF_ADS_TYPE = "adstype"
CONF_ADS_VALUE = "value"
CONF_ADS_VAR = "adsvar"
CONF_ADS_VAR_BRIGHTNESS = "adsvar_brightness"
CONF_ADS_VAR_POSITION = "adsvar_position"

STATE_KEY_STATE = "state"
STATE_KEY_BRIGHTNESS = "brightness"
STATE_KEY_POSITION = "position"

DOMAIN = "ads"

SERVICE_WRITE_DATA_BY_NAME = "write_data_by_name"

CONFIG_SCHEMA = vol.Schema(
    {
        DOMAIN: vol.Schema(
            {
                vol.Required(CONF_DEVICE): cv.string,
                vol.Required(CONF_PORT): cv.port,
                vol.Optional(CONF_IP_ADDRESS): cv.string,
            }
        )
    },
    extra=vol.ALLOW_EXTRA,
)

SCHEMA_SERVICE_WRITE_DATA_BY_NAME = vol.Schema(
    {
        vol.Required(CONF_ADS_TYPE): vol.In(
            [
                ADSTYPE_INT,
                ADSTYPE_UINT,
                ADSTYPE_BYTE,
                ADSTYPE_BOOL,
                ADSTYPE_DINT,
                ADSTYPE_UDINT,
            ]
        ),
        vol.Required(CONF_ADS_VALUE): vol.Coerce(int),
        vol.Required(CONF_ADS_VAR): cv.string,
    }
)


async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
    """Set up the ADS component."""

    conf = config[DOMAIN]

    net_id = conf[CONF_DEVICE]
    ip_address = conf.get(CONF_IP_ADDRESS)
    port = conf[CONF_PORT]

    client = pyads.Connection(net_id, port, ip_address)

    try:
        ads = AdsHub(client)
    except pyads.ADSError:
        _LOGGER.error(
            "Could not connect to ADS host (netid=%s, ip=%s, port=%s)",
            net_id,
            ip_address,
            port,
        )
        return False

    hass.data[DATA_ADS] = ads
    hass.async_create_background_task(ads.heartbeat(), "heartbeat")
    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, ads.shutdown)

    def handle_write_data_by_name(call: ServiceCall) -> None:
        """Write a value to the connected ADS device."""
        ads_var = call.data[CONF_ADS_VAR]
        ads_type = call.data[CONF_ADS_TYPE]
        value = call.data[CONF_ADS_VALUE]

        try:
            ads.write_by_name(ads_var, value, ADS_TYPEMAP[ads_type])
        except pyads.ADSError as err:
            _LOGGER.error(err)

    hass.services.async_register(
        DOMAIN,
        SERVICE_WRITE_DATA_BY_NAME,
        handle_write_data_by_name,
        schema=SCHEMA_SERVICE_WRITE_DATA_BY_NAME,
    )

    return True


# Tuple to hold data needed for notification
NotificationItem = namedtuple(
    "NotificationItem", "hnotify huser name plc_datatype callback"
)

# Tuple to hold data needed to restore notification
DeletedNotification = namedtuple("DeletedNotification", "name plc_datatype callback")


class ConnectionState(Enum):
    """Reresentation of ADS connection state."""

    Connected = 1
    ReadyToReconnect = 2
    Disconnected = 3


class AdsHub:
    """Representation of an ADS connection."""

    def __init__(self, ads_client: pyads.Connection) -> None:
        """Initialize the ADS hub."""
        self._client = ads_client
        self._client.open()
        self._is_running = True

        # All ADS devices are registered here
        self._notification_items: dict[int, NotificationItem] = {}
        self._lock = threading.Lock()

    def shutdown(self, *args, **kwargs):
        """Shutdown ADS connection."""

        _LOGGER.debug("Shutting down ADS")

        self._is_running = False
        self._delete_device_notifications()

        try:
            self._client.close()
        except pyads.ADSError as err:
            _LOGGER.error(err)

    def write_by_name(self, name, value, plc_datatype):
        """Write a value to the device."""

        with self._lock:
            try:
                return self._client.write_by_name(name, value, plc_datatype)
            except pyads.ADSError as err:
                _LOGGER.error("Error writing %s: %s", name, err)

    def read_by_name(self, name, plc_datatype):
        """Read a value from the device."""

        with self._lock:
            try:
                return self._client.read_by_name(name, plc_datatype)
            except pyads.ADSError as err:
                _LOGGER.error("Error reading %s: %s", name, err)

    def add_device_notification(self, name, plc_datatype, callback):
        """Add a notification to the ADS devices."""

        attr = pyads.NotificationAttrib(ctypes.sizeof(plc_datatype))

        with self._lock:
            try:
                hnotify, huser = self._client.add_device_notification(
                    name, attr, self._device_notification_callback
                )
            except pyads.ADSError as err:
                _LOGGER.error("Error subscribing to %s: %s", name, err)
            else:
                hnotify = int(hnotify)
                self._notification_items[hnotify] = NotificationItem(
                    hnotify, huser, name, plc_datatype, callback
                )

                _LOGGER.debug(
                    "Added device notification %d for variable %s", hnotify, name
                )

    async def heartbeat(
        self,
        heartbeat_interval: float = 5.0,
        max_wait_time: float = 120.0,
        communication_timeout_while_disconnected_ms: int = 100,
    ):
        """Periodically checks and handles the connection state of the client.

        Adjusts the wait time and communication timeout based on the connection state, with exponential backoff when disconnected.

        Attempts reconnection and manages device notifications upon reconnection.
        """
        deleted_device_notifications: list[DeletedNotification] = []

        default_timeout_ms = 5000

        wait_time = heartbeat_interval

        while self._is_running:
            connection_state = self._check_connection()
            if connection_state == ConnectionState.Connected:
                if wait_time != heartbeat_interval:
                    _LOGGER.info("Reconnected")
                    self._client.set_timeout(ms=default_timeout_ms)
                    wait_time = heartbeat_interval

                await asyncio.sleep(wait_time)
            elif connection_state == ConnectionState.Disconnected:
                if wait_time == heartbeat_interval:
                    _LOGGER.info("Disconnected, waiting for device availability")
                    self._client.set_timeout(
                        ms=communication_timeout_while_disconnected_ms
                    )

                wait_time = (
                    wait_time * 2 if wait_time * 2 < max_wait_time else max_wait_time
                )
                _LOGGER.debug("Waiting %d seconds to check device state", wait_time)
                await asyncio.sleep(wait_time)
            elif connection_state == ConnectionState.ReadyToReconnect:
                if not deleted_device_notifications:
                    deleted_device_notifications = self._delete_device_notifications()

                self._reconnect()

                while deleted_device_notifications:
                    self.add_device_notification(*deleted_device_notifications.pop())

    def _check_connection(self) -> ConnectionState:
        try:
            self._client.read_state()
        except pyads.ADSError as read_state_error:
            if read_state_error.err_code not in pyads.errorcodes.ERROR_CODES:
                return ConnectionState.ReadyToReconnect
            return ConnectionState.Disconnected
        else:
            return ConnectionState.Connected

    def _reconnect(self):
        try:
            self._client.close()
            self._client.open()
        except pyads.ADSError as reconnect_error:
            _LOGGER.error(reconnect_error)
            return

    def _device_notification_callback(self, notification, name):
        """Handle device notifications."""
        contents = notification.contents

        hnotify = int(contents.hNotification)
        _LOGGER.debug("Received notification %d", hnotify)

        # get dynamically sized data array
        data_size = contents.cbSampleSize
        data = (ctypes.c_ubyte * data_size).from_address(
            ctypes.addressof(contents)
            + pyads.structs.SAdsNotificationHeader.data.offset
        )

        try:
            with self._lock:
                notification_item = self._notification_items[hnotify]
        except KeyError:
            _LOGGER.error("Unknown device notification handle: %d", hnotify)
            return

        # Parse data to desired datatype
        if notification_item.plc_datatype == pyads.PLCTYPE_BOOL:
            value = bool(struct.unpack("<?", bytearray(data))[0])
        elif notification_item.plc_datatype == pyads.PLCTYPE_INT:
            value = struct.unpack("<h", bytearray(data))[0]
        elif notification_item.plc_datatype == pyads.PLCTYPE_BYTE:
            value = struct.unpack("<B", bytearray(data))[0]
        elif notification_item.plc_datatype == pyads.PLCTYPE_UINT:
            value = struct.unpack("<H", bytearray(data))[0]
        elif notification_item.plc_datatype == pyads.PLCTYPE_DINT:
            value = struct.unpack("<i", bytearray(data))[0]
        elif notification_item.plc_datatype == pyads.PLCTYPE_UDINT:
            value = struct.unpack("<I", bytearray(data))[0]
        else:
            value = bytearray(data)
            _LOGGER.warning("No callback available for this datatype")

        notification_item.callback(notification_item.name, value)

    def _delete_device_notifications(self) -> list[DeletedNotification]:
        result = [
            DeletedNotification(item.name, item.plc_datatype, item.callback)
            for item in self._notification_items.values()
        ]

        while self._notification_items:
            _, notification_item = self._notification_items.popitem()
            _LOGGER.debug(
                "Deleting device notification %d, %d",
                notification_item.hnotify,
                notification_item.huser,
            )
            try:
                self._client.del_device_notification(
                    notification_item.hnotify, notification_item.huser
                )
            except pyads.ADSError as err:
                _LOGGER.error(err)

        return result


class AdsEntity(Entity):
    """Representation of ADS entity."""

    _attr_should_poll = False

    def __init__(self, ads_hub, name, ads_var) -> None:
        """Initialize ADS binary sensor."""
        self._state_dict: dict[str, Any] = {}
        self._state_dict[STATE_KEY_STATE] = None
        self._ads_hub = ads_hub
        self._ads_var = ads_var
        self._event = None
        self._attr_unique_id = ads_var
        self._attr_name = name

    async def async_initialize_device(
        self, ads_var, plctype, state_key=STATE_KEY_STATE, factor=None
    ):
        """Register device notification."""

        def update(name, value):
            """Handle device notifications."""
            _LOGGER.debug("Variable %s changed its value to %d", name, value)

            if factor is None:
                self._state_dict[state_key] = value
            else:
                self._state_dict[state_key] = value / factor

            asyncio.run_coroutine_threadsafe(async_event_set(), self.hass.loop)
            self.schedule_update_ha_state()

        async def async_event_set():
            """Set event in async context."""
            self._event.set()

        self._event = asyncio.Event()

        await self.hass.async_add_executor_job(
            self._ads_hub.add_device_notification, ads_var, plctype, update
        )
        try:
            async with timeout(10):
                await self._event.wait()
        except TimeoutError:
            _LOGGER.debug("Variable %s: Timeout during first update", ads_var)

    @property
    def available(self) -> bool:
        """Return False if state has not been updated yet."""
        return self._state_dict[STATE_KEY_STATE] is not None
