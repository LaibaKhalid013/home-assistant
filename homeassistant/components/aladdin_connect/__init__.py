"""The Aladdin Connect Genie integration."""

from __future__ import annotations

from typing import TYPE_CHECKING

from homeassistant.config_entries import ConfigEntry, ConfigEntryState
from homeassistant.helpers import issue_registry as ir

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant

DOMAIN = "aladdin_connect"


async def async_setup_entry(hass: HomeAssistant, _: ConfigEntry) -> bool:
    """Set up Aladdin Connect from a config entry."""
    ir.async_create_issue(
        hass,
        DOMAIN,
        DOMAIN,
        is_fixable=False,
        severity=ir.IssueSeverity.ERROR,
        translation_key="integration_removed",
        translation_placeholders={
            "entries": "/config/integrations/integration/aladdin_connect",
        },
    )

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    if all(
        config_entry.state is ConfigEntryState.NOT_LOADED
        for config_entry in hass.config_entries.async_entries(DOMAIN)
        if config_entry.entry_id != entry.entry_id
    ):
        ir.async_delete_issue(hass, DOMAIN, DOMAIN)

    return True
