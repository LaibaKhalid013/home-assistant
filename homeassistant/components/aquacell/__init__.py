"""The Aquacell integration."""
from __future__ import annotations

from aioaquacell import AquacellApi

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import DATA_AQUACELL, DOMAIN
from .coordinator import AquacellCoordinator

PLATFORMS = [Platform.SENSOR]


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Aquacell from a config entry."""
    session = async_get_clientsession(hass)

    aquacell_api = AquacellApi(session)

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {DATA_AQUACELL: aquacell_api}

    coordinator = AquacellCoordinator(hass, aquacell_api, entry)

    await coordinator.async_config_entry_first_refresh()
    hass.data[DOMAIN][entry.entry_id] = coordinator

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        hass.data[DOMAIN].pop(entry.entry_id)

    return unload_ok
