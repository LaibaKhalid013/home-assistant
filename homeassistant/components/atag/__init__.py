"""The ATAG Integration."""

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant

from .coordinator import AtagDataUpdateCoordinator

DOMAIN = "atag"
PLATFORMS = [Platform.CLIMATE, Platform.SENSOR, Platform.WATER_HEATER]


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Atag integration from a config entry."""

    coordinator = AtagDataUpdateCoordinator(hass, entry)
    await coordinator.async_config_entry_first_refresh()

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
    if entry.unique_id is None:
        hass.config_entries.async_update_entry(entry, unique_id=coordinator.atag.id)

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload Atag config entry."""
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)

    if unload_ok:
        hass.data[DOMAIN].pop(entry.entry_id)
    return unload_ok
