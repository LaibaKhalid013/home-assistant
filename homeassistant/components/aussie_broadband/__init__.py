"""The Aussie Broadband integration."""

from __future__ import annotations

from aiohttp import ClientError
from aussiebb.asyncio import AussieBB
from aussiebb.const import FETCH_TYPES
from aussiebb.exceptions import AuthenticationException

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_PASSWORD, CONF_USERNAME, Platform
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import DOMAIN
from .coordinator import AussieBroadandDataUpdateCoordinator

PLATFORMS = [Platform.SENSOR]


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up Aussie Broadband from a config entry."""
    # Login to the Aussie Broadband API and retrieve the current service list
    client = AussieBB(
        entry.data[CONF_USERNAME],
        entry.data[CONF_PASSWORD],
        async_get_clientsession(hass),
    )

    # Ignore services that don't support usage data
    ignore_types = [*FETCH_TYPES, "Hardware"]

    try:
        await client.login()
        services = await client.get_services(drop_types=ignore_types)
    except AuthenticationException as exc:
        raise ConfigEntryAuthFailed from exc
    except ClientError as exc:
        raise ConfigEntryNotReady from exc

    # Initiate a Data Update Coordinator for each service
    for service in services:
        service["coordinator"] = AussieBroadandDataUpdateCoordinator(
            hass, client, service["service_id"]
        )
        await service["coordinator"].async_config_entry_first_refresh()

    # Setup the integration
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
        "client": client,
        "services": services,
    }
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload the config entry."""
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        hass.data[DOMAIN].pop(entry.entry_id)

    return unload_ok
