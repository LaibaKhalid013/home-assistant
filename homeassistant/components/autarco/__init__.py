"""The Autarco integration."""

from __future__ import annotations

from autarco import Autarco

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_EMAIL, CONF_PASSWORD, Platform
from homeassistant.core import HomeAssistant
from homeassistant.helpers.aiohttp_client import async_get_clientsession

from .const import DOMAIN
from .coordinator import AutarcoDataUpdateCoordinator

PLATFORMS: list[Platform] = [Platform.SENSOR]

type AutarcoConfigEntry = ConfigEntry[AutarcoDataUpdateCoordinator]


async def async_setup_entry(hass: HomeAssistant, entry: AutarcoConfigEntry) -> bool:
    """Set up Autarco from a config entry."""
    client = Autarco(
        email=entry.data[CONF_EMAIL],
        password=entry.data[CONF_PASSWORD],
        session=async_get_clientsession(hass),
    )

    coordinator = AutarcoDataUpdateCoordinator(hass, client)
    await coordinator.async_config_entry_first_refresh()

    hass.data.setdefault(DOMAIN, {})
    hass.data[DOMAIN][entry.entry_id] = coordinator

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        del hass.data[DOMAIN][entry.entry_id]
    return unload_ok
