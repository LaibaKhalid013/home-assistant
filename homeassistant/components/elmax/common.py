"""Elmax integration common classes and utilities."""

from __future__ import annotations

import ssl

from elmax_api.model.endpoint import DeviceEndpoint
from elmax_api.model.panel import PanelEntry
from packaging import version

from homeassistant.core import callback
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.dispatcher import async_dispatcher_connect
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import (
    DOMAIN,
    ELMAX_LOCAL_API_PATH,
    MIN_APIV2_SUPPORTED_VERSION,
    SIGNAL_PANEL_UPDATE,
)
from .coordinator import ElmaxCoordinator


def get_direct_api_url(host: str, port: int, use_ssl: bool) -> str:
    """Return the direct API url given the base URI."""
    schema = "https" if use_ssl else "http"
    return f"{schema}://{host}:{port}/{ELMAX_LOCAL_API_PATH}"


def build_direct_ssl_context(cadata: str) -> ssl.SSLContext:
    """Create a custom SSL context for direct-api verification."""
    context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLS_CLIENT)
    context.check_hostname = False
    context.verify_mode = ssl.CERT_REQUIRED
    context.load_verify_locations(cadata=cadata)
    return context


def check_local_version_supported(api_version: str | None) -> bool:
    """Check whether the given API version is supported."""
    if api_version is None:
        return False
    return version.parse(api_version) >= version.parse(MIN_APIV2_SUPPORTED_VERSION)


class DirectPanel(PanelEntry):
    """Helper class for wrapping a directly accessed Elmax Panel."""

    def __init__(self, panel_uri):
        """Construct the object."""
        super().__init__(panel_uri, True, {})

    def get_name_by_user(self, username: str) -> str:
        """Return the panel name."""
        return f"Direct Panel {self.hash}"


class ElmaxEntity(CoordinatorEntity[ElmaxCoordinator]):
    """Wrapper for Elmax entities."""

    _last_state: DeviceEndpoint = None

    def __init__(
        self,
        elmax_device: DeviceEndpoint,
        panel_version: str,
        coordinator: ElmaxCoordinator,
    ) -> None:
        """Construct the object."""
        super().__init__(coordinator=coordinator)
        self._device = elmax_device
        self._attr_unique_id = elmax_device.endpoint_id
        self._attr_name = elmax_device.name
        self._attr_device_info = DeviceInfo(
            identifiers={(DOMAIN, coordinator.panel_entry.hash)},
            name=coordinator.panel_entry.get_name_by_user(
                coordinator.http_client.get_authenticated_username()
            ),
            manufacturer="Elmax",
            model=panel_version,
            sw_version=panel_version,
        )

    async def async_added_to_hass(self) -> None:
        """Register push notifications callbacks if available."""
        self.async_on_remove(
            async_dispatcher_connect(
                self.hass,
                f"{SIGNAL_PANEL_UPDATE}-{self.coordinator.panel_entry.hash}-{self._device.endpoint_id}",
                self._handle_update,
            )
        )
        return await super().async_added_to_hass()

    @callback
    def _handle_update(self, endpoint_status: DeviceEndpoint) -> None:
        self._last_state = endpoint_status
        self.async_write_ha_state()

    @property
    def available(self) -> bool:
        """Return if entity is available."""
        return super().available and self.coordinator.panel_entry.online
