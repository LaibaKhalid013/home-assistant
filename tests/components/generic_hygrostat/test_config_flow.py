"""Test the generic hygrostat config flow."""

from unittest.mock import patch

from syrupy.assertion import SnapshotAssertion
from syrupy.filters import props

from homeassistant.components.generic_hygrostat import (
    CONF_AWAY_HUMIDITY,
    CONF_DEVICE_CLASS,
    CONF_DRY_TOLERANCE,
    CONF_HUMIDIFIER,
    CONF_INITIAL_STATE,
    CONF_NAME,
    CONF_SENSOR,
    CONF_WET_TOLERANCE,
    DOMAIN,
)
from homeassistant.config_entries import SOURCE_USER
from homeassistant.core import HomeAssistant

from tests.common import MockConfigEntry, async_mock_service


async def test_config_flow(hass: HomeAssistant, snapshot: SnapshotAssertion) -> None:
    """Test the config flow."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": SOURCE_USER}
    )
    assert result == snapshot(name="init", exclude=props("data_schema"))

    with patch(
        "homeassistant.components.generic_hygrostat.async_setup_entry",
        return_value=True,
    ) as mock_setup_entry:
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_NAME: "My dehumidifier",
                CONF_DRY_TOLERANCE: 2,
                CONF_WET_TOLERANCE: 4,
                CONF_HUMIDIFIER: "switch.run",
                CONF_SENSOR: "sensor.humidity",
                CONF_DEVICE_CLASS: "dehumidifier",
                CONF_AWAY_HUMIDITY: 35,
                CONF_INITIAL_STATE: True,
            },
        )
        await hass.async_block_till_done()

    assert result == snapshot(name="create")
    assert len(mock_setup_entry.mock_calls) == 1

    config_entry = hass.config_entries.async_entries(DOMAIN)[0]
    assert config_entry.data == {}
    assert result == snapshot(name="options")
    assert config_entry.title == "My dehumidifier"


async def test_options(hass: HomeAssistant, snapshot: SnapshotAssertion) -> None:
    """Test reconfiguring."""

    turn_off_calls = async_mock_service(hass, "homeassistant", "turn_off")
    turn_on_calls = async_mock_service(hass, "homeassistant", "turn_on")

    config_entry = MockConfigEntry(
        data={},
        domain=DOMAIN,
        options={
            CONF_AWAY_HUMIDITY: 35.0,
            CONF_DEVICE_CLASS: "dehumidifier",
            CONF_DRY_TOLERANCE: 2.0,
            CONF_HUMIDIFIER: "switch.run",
            CONF_INITIAL_STATE: True,
            CONF_NAME: "My dehumidifier",
            CONF_SENSOR: "sensor.humidity",
            CONF_WET_TOLERANCE: 4.0,
        },
        title="My dehumidifier",
    )
    config_entry.add_to_hass(hass)

    # start with a humidity less than max, and a switch that is on
    hass.states.async_set(
        "sensor.humidity",
        "10",
        {"unit_of_measurement": "%", "device_class": "humidity"},
    )
    hass.states.async_set("switch.run", "on")

    assert await hass.config_entries.async_setup(config_entry.entry_id)
    await hass.async_block_till_done()

    # will turn off on start, since humidity is less than max
    assert len(turn_on_calls) == 0
    assert len(turn_off_calls) == 1

    result = await hass.config_entries.options.async_init(config_entry.entry_id)
    assert result == snapshot(name="init", exclude=props("data_schema", "handler"))

    # check that it is setup
    await hass.async_block_till_done()
    assert hass.states.get("humidifier.my_dehumidifier") == snapshot(name="with_away")

    # remove away feature
    result = await hass.config_entries.options.async_configure(
        result["flow_id"],
        user_input={
            CONF_DRY_TOLERANCE: 2,
            CONF_WET_TOLERANCE: 4,
            CONF_HUMIDIFIER: "switch.run",
            CONF_SENSOR: "sensor.humidity",
            CONF_DEVICE_CLASS: "dehumidifier",
            CONF_INITIAL_STATE: True,
        },
    )
    assert result == snapshot(name="create_entry", exclude=props("handler"))

    # Check config entry is reloaded with new options
    await hass.async_block_till_done()
    assert hass.states.get("humidifier.my_dehumidifier") == snapshot(
        name="without_away"
    )
