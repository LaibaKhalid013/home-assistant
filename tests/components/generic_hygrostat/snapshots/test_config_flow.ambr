# serializer version: 1
# name: test_config_flow[create]
  FlowResultSnapshot({
    'context': dict({
      'source': 'user',
    }),
    'data': dict({
    }),
    'description': None,
    'description_placeholders': None,
    'flow_id': <ANY>,
    'handler': 'generic_hygrostat',
    'minor_version': 1,
    'options': dict({
      'away_humidity': 35.0,
      'device_class': 'dehumidifier',
      'dry_tolerance': 2.0,
      'humidifier': 'switch.run',
      'initial_state': 1.0,
      'name': 'My dehumidifier',
      'target_sensor': 'sensor.humidity',
      'wet_tolerance': 4.0,
    }),
    'result': ConfigEntrySnapshot({
      'data': dict({
      }),
      'disabled_by': None,
      'domain': 'generic_hygrostat',
      'entry_id': <ANY>,
      'minor_version': 1,
      'options': dict({
        'away_humidity': 35.0,
        'device_class': 'dehumidifier',
        'dry_tolerance': 2.0,
        'humidifier': 'switch.run',
        'initial_state': 1.0,
        'name': 'My dehumidifier',
        'target_sensor': 'sensor.humidity',
        'wet_tolerance': 4.0,
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'title': 'My dehumidifier',
      'unique_id': None,
      'version': 1,
    }),
    'title': 'My dehumidifier',
    'type': <FlowResultType.CREATE_ENTRY: 'create_entry'>,
    'version': 1,
  })
# ---
# name: test_config_flow[init]
  FlowResultSnapshot({
    'description_placeholders': None,
    'errors': None,
    'flow_id': <ANY>,
    'handler': 'generic_hygrostat',
    'last_step': True,
    'preview': None,
    'step_id': 'user',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_config_flow[options]
  FlowResultSnapshot({
    'context': dict({
      'source': 'user',
    }),
    'data': dict({
    }),
    'description': None,
    'description_placeholders': None,
    'flow_id': <ANY>,
    'handler': 'generic_hygrostat',
    'minor_version': 1,
    'options': dict({
      'away_humidity': 35.0,
      'device_class': 'dehumidifier',
      'dry_tolerance': 2.0,
      'humidifier': 'switch.run',
      'initial_state': 1.0,
      'name': 'My dehumidifier',
      'target_sensor': 'sensor.humidity',
      'wet_tolerance': 4.0,
    }),
    'result': ConfigEntrySnapshot({
      'data': dict({
      }),
      'disabled_by': None,
      'domain': 'generic_hygrostat',
      'entry_id': <ANY>,
      'minor_version': 1,
      'options': dict({
        'away_humidity': 35.0,
        'device_class': 'dehumidifier',
        'dry_tolerance': 2.0,
        'humidifier': 'switch.run',
        'initial_state': 1.0,
        'name': 'My dehumidifier',
        'target_sensor': 'sensor.humidity',
        'wet_tolerance': 4.0,
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'title': 'My dehumidifier',
      'unique_id': None,
      'version': 1,
    }),
    'title': 'My dehumidifier',
    'type': <FlowResultType.CREATE_ENTRY: 'create_entry'>,
    'version': 1,
  })
# ---
# name: test_options[create_entry]
  FlowResultSnapshot({
    'context': dict({
    }),
    'data': dict({
      'device_class': 'dehumidifier',
      'dry_tolerance': 2.0,
      'humidifier': 'switch.run',
      'initial_state': 1.0,
      'name': 'My dehumidifier',
      'target_sensor': 'sensor.humidity',
      'wet_tolerance': 4.0,
    }),
    'description': None,
    'description_placeholders': None,
    'flow_id': <ANY>,
    'result': True,
    'type': <FlowResultType.CREATE_ENTRY: 'create_entry'>,
  })
# ---
# name: test_options[init]
  FlowResultSnapshot({
    'description_placeholders': None,
    'errors': None,
    'flow_id': <ANY>,
    'last_step': True,
    'preview': None,
    'step_id': 'init',
    'type': <FlowResultType.FORM: 'form'>,
  })
# ---
# name: test_options[options]
  dict({
    'device_class': 'dehumidifier',
    'dry_tolerance': 2.0,
    'humidifier': 'switch.run',
    'initial_state': 1.0,
    'name': 'My dehumidifier',
    'target_sensor': 'sensor.humidity',
    'wet_tolerance': 4.0,
  })
# ---
# name: test_options[with_away]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'action': <HumidifierAction.IDLE: 'idle'>,
      'available_modes': list([
        'normal',
        'away',
      ]),
      'current_humidity': 10.0,
      'device_class': 'dehumidifier',
      'friendly_name': 'My dehumidifier',
      'humidity': 100,
      'max_humidity': 100,
      'min_humidity': 0,
      'mode': 'normal',
      'saved_humidity': 35.0,
      'supported_features': <HumidifierEntityFeature: 1>,
    }),
    'context': <ANY>,
    'entity_id': 'humidifier.my_dehumidifier',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
# name: test_options[without_away]
  StateSnapshot({
    'attributes': ReadOnlyDict({
      'action': <HumidifierAction.IDLE: 'idle'>,
      'current_humidity': 10.0,
      'device_class': 'dehumidifier',
      'friendly_name': 'My dehumidifier',
      'humidity': 100,
      'max_humidity': 100,
      'min_humidity': 0,
      'saved_humidity': 35,
      'supported_features': <HumidifierEntityFeature: 0>,
    }),
    'context': <ANY>,
    'entity_id': 'humidifier.my_dehumidifier',
    'last_changed': <ANY>,
    'last_reported': <ANY>,
    'last_updated': <ANY>,
    'state': 'on',
  })
# ---
